<a href="https://colab.research.google.com/github/bhuvansun/data-science-salaries/blob/main/assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> ### **Assignment 2 - Numpy Array Operations**



# 5 Interesting Numpy Functions


### That make Numpy so impressive


- Function 1: np.logspace()
- Function 2: np.moveaxis()
- Function 3: np.dstack()
- Function 4: np.bitwise_xor()
- Function 5: np.matmul()

In [ ]:
!pip install jovian --upgrade -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.6/68.6 kB 2.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


In [ ]:
import jovian

In [ ]:
jovian.commit(project='assignment-2')

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


In [ ]:
import numpy as np

In [ ]:
# List of functions explained
function1 = np.logspace()
function2 =
function3 =
function4 =
function5 =

## Function 1 - np.logspace()

Similar to linspace, logspace returns an array of evenly spaced numbers between specified numbers, but on a log scale with a particular base.

Calling method:
numpy.logspace(start, stop, num='(int)', endpoint='(Bool)', base='(float)', dtype=None, axis='(int)')

If not specified, the base is taken as 10 by default, and endpoint is taken as True by default.



In [ ]:
# Example 1 - working
np.logspace(2.0, 3.0, num=4, base=2.0)

array([4.        , 5.0396842 , 6.34960421, 8.        ])

Here, we select 4 numbers between 2 and 3, which would be 2, 2.33, 2.66, 3.


The base is specified as 2.


Now, the function generates an array which contains an ordered sequence of the values which, when the log is applied on them, return the above mentioned values.


For example, log(base=2)4 would give 2, which is the first element.


Similar is the case with the other elements.

In [ ]:
# Example 2 - working
np.logspace(2.0, 5.0, num=4, endpoint=False)

array([  100.        ,   562.34132519,  3162.27766017, 17782.79410039])

Here, we select 4 numbers between 2 and 5.

Since we set the endpoint to False, this selection is exclusive of 5.
Numpy selected 4.24 as the last number in this case.

The default base is 10.

The output array contains 4 numbers which give us the selected numbers when the log is applied on them.

Log of 562 is 2.74, 3162 is 3.49, and 17782 is 4.24.

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
np.logspace(-1, 10.0, num=4, base=-2.0)

/usr/local/lib/python3.10/dist-packages/numpy/core/function_base.py:277: RuntimeWarning: invalid value encountered in power
  return _nx.power(base, y)


array([-5.000e-01,        nan,        nan,  1.024e+03])

Over here, we mentioned the base to be -2.

Although the datatype is a float, base of a log function cannot be negative.

Hence, we get an error regarding the invalid value of the base.

Usually, when representing large values in a graph, the values are represented in a logarithmic scale with some base(typically 10) for convenience, as displaying on a linear scale won't give a proper visual of the data due to thee massice size.

One best example is the representation of the distance of planets and celestial bodies from the earth. A log scale would give a cleaner representation, whereas a linear scale would be too large to even fit into a graph.

Using logspace, we can derive the linear values between any two endpoints quickly for deriving the actual value from the log scale.

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


## Function 2 - np.moveaxis()

Moves the axes of an array according to the parameters specified.


Function calling:
numpy.moveaxis(a, source, destination)

------

source:


Original positions of the axes to move. These must be unique.

destination:


Destination positions for each of the original axes. These must also be unique.


In [ ]:
# Example 1 - working
x = np.zeros((3, 4, 5))
np.moveaxis(x, 0, -1).shape

(4, 5, 3)

The given array contains 3 2D arrays with 4 1D arrays with 5 elements in each 1D array.

After altering the axes, the new array has 4 2D arrays with 5 1D arrays with 3 elements in each 1D array.

In [ ]:
# Example 2 - working
np.moveaxis(x, -1, 0).shape

(5, 3, 4)

The given array contains 3 2D arrays with 4 1D arrays with 5 elements in each 1D array.

After altering the axes, the new array has 5 2D arrays with 3 1D arrays with 4 elements in each 1D array.

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
np.moveaxis(x, -1, 3).shape

AxisError: ignored

The range of the axes should be within the dimensions of the array. In this case, it should be within 3.

This function can be used when we need to rotate the array for specific operations such as multiplication etc.

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


## Function 3 - np.dstack()

Function for stacking arrays along the third axis(depth wise).

Function calling:

numpy.dstack('array')

In [ ]:
# Example 1 - working
a = np.ones([4, 5])
b = 2 * np.ones([4, 5])

np.dstack([a, b])

array([[[1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.]],

       [[1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.]],

       [[1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.]],

       [[1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.],
        [1., 2.]]])

a is an array with ones arrange in a 2D array of size 4*5.

b is similar to a, but with twos arranged in the same fashion.

Now, when we use dstack, these two arrays are stacked along the third axis, making them into a 3D array, with the values arranged sequentially as 1, 2.

In [ ]:
# Example 2 - working
a = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([[7, 8, 9], [10, 11, 12]])

np.dstack([a, b])

array([[[ 1,  7],
        [ 2,  8],
        [ 3,  9]],

       [[ 4, 10],
        [ 5, 11],
        [ 6, 12]]])

a and b are two arrays of dimension 2*3.


When we depth stack the arrays, we obtain a 3D array with the values of the second array placed adjacent to the first array.

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.ones([4, 6])
b = 2 * np.ones([4, 5])

np.dstack([a, b])

ValueError: ignored

Similar to other array operations involving two arrays, the size of the arrays should be the same to depth stack them against each other.

Useful for combining two arrays in a non-linear sense.

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


## Function 4 - numpy.bitwise_xor()

Function for computing bit-wise XOR of two arrays.


Function calling:


numpy.bitwise_xor(array1, array2)

In [ ]:
# Example 1 - working
a = np.array([[1, 3], [2, 4]])
b = np.array([[5, 6], [7, 8]])

np.bitwise_xor(a, b)

array([[ 4,  5],
       [ 5, 12]])

XOR operation between two bits gives a 1 if the bits are different, and gives a 0 if the bits are the same.


Arrays a and b are 2*2 sized 2D arrays. The operation happens element-wise, meaning, it happens between a[0][0] and b[0][0], a[0][1] and b[0][1] and so on.


For example, we have a[0][0] = 1 = 001 and b[0][0] = 5 = 101. XOR operation gives 100 = 4.


The same process repeated for all the elements gives an array [4, 5, 5, 12] as the result.

In [ ]:
# Example 2 - working
x1 = np.array([True, True])
x2 = np.array([False, True])
x1 ^ x2

array([ True, False])

XOR operation between True and False gives True, since the boolean values are different. Similarly, we get a False if both the Boolean values are the same.


Hence, the result in this case would be [True, False]

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
a = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([[2, 5], [4, 7]])

np.bitwise_xor(a, b)

ValueError: ignored

Since the XOR operation happens element-wise, the size of the arrays should be the same.

This function can be used to save time and effort wherever we need to perform bit operations, in this case, XOR operation.

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


## Function 5 - numpy.matmul()

Used for matrix multiplication between two matrices.


Function Calling: np.matmul(array1, array2)

In [ ]:
# Example 1 - working
a = np.array([[1, 0],
              [0, 1]])
b = np.array([[4, 1],
              [2, 2]])
np.matmul(a, b)

array([[4, 1],
       [2, 2]])

For the above case, the first element[0][0] of the matrix multiplication would be the sum of the product of (1, 4) and (0, 2), which is 4.


Similarly, [0][1] would be sum of (1, 1) and (0, 2), which is 1.


Hence, the final array would be [[4, 1], [2, 2]].

In [ ]:
# Example 2 - working
a = np.array([[1, 0],
              [0, 1]])
b = np.array([1, 2])

np.matmul(a, b)

array([1, 2])

Matrix multiplication can be done on 2 arrays if the column size of the first array is equal to the row size of the second array, which is satisifed in the above case of multiplication between a 2D and a 1D array.

In [ ]:
# Example 3 - breaking (to illustrate when it breaks)
np.matmul([1,2], 3)

ValueError: ignored

Normal multiplication with a scalar is not possible in matrix multiplication. One solution wuld be to broadcast the scalar to an array.

Useful for performing matrix multiplication between two arrays without needing to write any code for doing the same.

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.


## Conclusion

We have learnt the use of numpy library, and the features of many of its functions

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* Numpy GeeksforGeeks: https://www.geeksforgeeks.org/python-numpy/

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] jovian.commit() is no longer required on Google Colab. If you ran this notebook from Jovian, 
then just save this file in Colab using Ctrl+S/Cmd+S and it will be updated on Jovian. 
Also, you can also delete this cell, it's no longer necessary.
